In [1]:
install.packages("kernlab")
library(kernlab)


  There is a binary version available but the source version is later:
        binary source needs_compilation
kernlab 0.9-29 0.9-32              TRUE

  Binaries will be installed
package 'kernlab' successfully unpacked and MD5 sums checked


Warning message:
"cannot remove prior installation of package 'kernlab'"Warning message in file.copy(savedcopy, lib, recursive = TRUE):
"problem copying C:\Users\Tien\anaconda3\envs\new evn\Lib\R\library\00LOCK\kernlab\libs\x64\kernlab.dll to C:\Users\Tien\anaconda3\envs\new evn\Lib\R\library\kernlab\libs\x64\kernlab.dll: Permission denied"Warning message:
"restored 'kernlab'"


The downloaded binary packages are in
	C:\Users\Tien\AppData\Local\Temp\RtmpU5lUTd\downloaded_packages


# Question 1

In [2]:
data <- read.table("./data 2.2/credit_card_data.txt") 
head(data)

V1 V2    V3    V4   V5 V6 V7 V8 V9  V10 V11
1 1  30.83 0.000 1.25 1  0  1  1  202   0 1  
2 0  58.67 4.460 3.04 1  0  6  1   43 560 1  
3 0  24.50 0.500 1.50 1  1  0  1  280 824 1  
4 1  27.83 1.540 3.75 1  0  5  0  100   3 1  
5 1  20.17 5.625 1.71 1  1  0  1  120   0 1  
6 1  32.08 4.000 2.50 1  1  0  0  360   0 1

In [3]:
set.seed(1)

In [4]:
model <- ksvm(as.matrix(data[,1:10]),as.factor(data[,11]), C=50, scaled = T, kernel="vanilladot", type = "C-svc")

 Setting default kernel parameters  


In [5]:
a <- colSums(model@xmatrix[[1]] * model@coef[[1]])
a

V1            V2            V3            V4            V5 
-0.0010523630 -0.0012025131 -0.0015382662  0.0028761998  1.0052764944 
           V6            V7            V8            V9           V10 
-0.0024958086  0.0001810245 -0.0006514829 -0.0013757143  0.1064002847

In [6]:
a0 <- -model@b
a0

[1] 0.08147145

The classifier's equation is then: -0.0010523630v1 -0.0012025131v2 -0.0015382662v3 + 0.0028761998v4 + 1.0052764944v5 -0.0024958086v6 + 0.0001810245v7 -0.0006514829v8 -0.0013757143v9 + 0.1064002847v10 + 0.08147145 = 0

In [7]:
pred <- predict(model,data[,1:10])

In [8]:
round(sum(pred == data[,11]) / nrow(data) * 100 , 2)

[1] 86.39

### The model's accuracy is 86.39%

# Question 2

In [9]:
library(kknn)

Warning message:
"package 'kknn' was built under R version 3.6.3"

In [10]:
check_accuracy = function(X){
  predict <- rep(0,(nrow(data)))
  for (i in 1:nrow(data)){
    model=kknn(V11~V1+V2+V3+V4+V5+V6+V7+V8+V9+V10,data[-i,],data[i,],k=X, scale = TRUE) 
    predict[i] <- as.integer(fitted(model)+0.5) 
  }
  accuracy = sum(predict == data[,11]) / nrow(data)
  return(accuracy)
}

In [11]:
accuracy <- rep(0,30) 
for (X in 1:20){
  accuracy[X] = check_accuracy(X) 
}

In [12]:
accuracy

[1] 0.8149847 0.8149847 0.8149847 0.8149847 0.8516820 0.8455657 0.8470948
 [8] 0.8486239 0.8470948 0.8501529 0.8516820 0.8532110 0.8516820 0.8516820
[15] 0.8532110 0.8516820 0.8516820 0.8516820 0.8501529 0.8501529 0.0000000
[22] 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000
[29] 0.0000000 0.0000000

### k = 12 and k = 15 are best accuracies because they have the biggest values (~0.853)